tests/end2end/fixtures/qutewm_sub.py

```diff
@@ -0,0 +1,400 @@
+        windows: A list of all managed windows in mapping order.
+        window_stack: A list of all windows in stack order.
+        root: The root window.
+        support_window: The window for the _NET_SUPPORTING_WM_CHECK.
+        atoms: AtomBag for the window manager.
+    """
+
+    WM_NAME = b'qutewm'
+
+    ROOT_EVENT_MASK = X.SubstructureNotifyMask | X.SubstructureRedirectMask
+    CLIENT_EVENT_MASK = (X.StructureNotifyMask | X.PropertyChangeMask |
+                         X.FocusChangeMask)
+
+    STATE_REMOVE = 0
+    STATE_ADD = 1
+    STATE_TOGGLE = 2
+
+    def __init__(self):
+        log.debug("initializing")
+        self._handlers = {
+            X.MapNotify: self._on_map_notify,
+            X.UnmapNotify: self._on_unmap_notify,
+            X.ClientMessage: self._on_client_message,
+            X.MapRequest: self._on_map_request,
+            X.ConfigureRequest: self._on_configure_request,
+            X.CirculateRequest: self._on_circulate_request,
+            X.PropertyNotify: self._on_property_notify,
+            X.FocusIn: self._on_focus_in,
+        }
+        self.dpy = Display()
+        self.dpy.set_error_handler(self._error_handler)
+
+        screen_width = self.dpy.screen().width_in_pixels
+        screen_height = self.dpy.screen().height_in_pixels
+        self.dimensions = (screen_width, screen_height)
+
+        self._retcode = None
+        self._needs_update = False
+        self.root = self.dpy.screen().root
+        self.support_window = None
+        self.windows = []
+        self.window_stack = []
+
+        self.root.change_attributes(event_mask=self.ROOT_EVENT_MASK,
+                                    onerror=self._wm_running)
+
+        self.atoms = AtomBag(self.dpy)
+
+        self._set_supported_attribute()
+        self._set_supporting_wm_check()
+
+    def _wm_running(self, error, request):
+        """Called when another WM is already running."""
+        log.error("Another window manager is running, exiting")
+        # This is called async, which means we can't just raise an exception,
+        # we need to signal the main thread to stop.
+        self._retcode = 42
+
+    def _error_handler(self, error, request):
+        # This is called async, which means we can't just raise an exception,
+        # we need to signal the main thread to stop.
+        if self._retcode is None:
+            log.error(error)
+
+    def _set_supported_attribute(self):
+        """Set the _NET_SUPPORTED attribute on the root window."""
+        attributes = [
+            self.atoms.supported,
+            self.atoms.active_window,
+            self.atoms.client_list,
+            self.atoms.wm_state,
+        ]
+        self.root.change_property(
+            self.atoms.supported,
+            Xatom.ATOM,
+            32,
+            attributes,
+        )
+
+    def _set_supporting_wm_check(self):
+        """Create and set a window for _NET_SUPPORTING_WM_CHECK."""
```

Comment:
It's just a window that is specified by EWMH, to make sure the window manager supports EWMH. I can add a comment with the explanation, or at least a link to the [spec](https://specifications.freedesktop.org/wm-spec/wm-spec-latest.html#idm140200472693600).